File: chapter2/exercise_22/go/temp_handler.go
package main

import "fmt"

// this task is specific for C++, following code is just code snippet similar to task

type Scale int

const (
	Celcius Scale = iota
	Kelvin
	Fahrenheit
)

func (c Scale) String() string {
	switch c {
	case Celcius:
		return "Celcius"
	case Kelvin:
		return "Kelvin"
	case Fahrenheit:
		return "Fahrenheit"
	default:
		return "unknown"
	}
}

type Temperature struct {
	val   float32
	scale Scale
}

func (v *Temperature) String() string {
	return fmt.Sprintf("%f.2 %s", v.val, v.scale)
}

func (v *Temperature) convert(toScale Scale) {
	switch v.scale {
	case Celcius:
		if toScale == Fahrenheit {
			v.val = v.val*9.0/5.0 + 32.0
			v.scale = Fahrenheit
		} else if toScale == Kelvin {
			v.val = v.val + 273.15
			v.scale = Kelvin
		}
	case Fahrenheit:
		if toScale == Celcius {
			v.val = (v.val - 32.0) * 5.0 / 9.0
			v.scale = Celcius
		} else if toScale == Kelvin {
			v.val = (v.val-32.0)*5.0/9.0 + 273.15
			v.scale = Kelvin
		}
	case Kelvin:
		if toScale == Celcius {
			v.val = v.val - 273.15
			v.scale = Celcius
		} else if toScale == Fahrenheit {
			v.val = (v.val-273.15)*9.0/5.0 + 32.0
			v.scale = Fahrenheit
		}
	}
}

func main() {
	temp := Temperature{val: -10, scale: Celcius}
	temp.convert(Fahrenheit)
	fmt.Println(temp)

}

File: chapter2/exercise_22/cpp/temp_handler.cpp
#include <cstdint>
#include <iostream>
#include <ostream>

namespace Temperature
{

enum class Scale
{
    Celcius,
    Fahrenheit,
    Kelvin
};

template <Scale s> class Value
{
  public:
    constexpr Value(long double t) : _t(t) {};
    constexpr Value operator-() const { return Value(-_t); }
    constexpr explicit operator double() const { return _t; };

  private:
    double _t;

    friend std::ostream &operator<<(std::ostream &stream, const Value<s> &v)
    {
        if constexpr (s == Scale::Celcius)
        {
            stream << v._t << " Celcius";
        }
        else if constexpr (s == Scale::Kelvin)
        {
            stream << v._t << " Kelvin";
        }
        else if constexpr (s == Scale::Fahrenheit)
        {
            stream << v._t << " Fahrenheit";
        }
        return stream;
    }
};

template <Scale s> constexpr bool operator>(const Value<s> &t1, const Value<s> &t2)
{
    return t1 > t2;
}
template <Scale s> constexpr bool operator<(const Value<s> &t1, const Value<s> &t2)
{
    return t1 < t2;
}
template <Scale s> constexpr bool operator>=(const Value<s> &t1, const Value<s> &t2)
{
    return t1 >= t2;
}
template <Scale s> constexpr bool operator<=(const Value<s> &t1, const Value<s> &t2)
{
    return t1 <= t2;
}
template <Scale s> constexpr bool operator==(const Value<s> &t1, const Value<s> &t2)
{
    return t1 == t2;
}
template <Scale s> constexpr bool operator!=(const Value<s> &t1, const Value<s> &t2)
{
    return t1 != t2;
}

template <Scale From, Scale T> struct conversion
{
    static double convert(const double value) = delete;
};

// I am too lazy to implement all the conversion...
template <> struct conversion<Scale::Celcius, Scale::Fahrenheit>
{
    static double convert(const double value) { return value * 9.0 / 5.0 + 32.0; }
};

template <Scale To, Scale From> Value<To> cast_scale(const Value<From> &t)
{
    return Value<To>(conversion<From, To>::convert(static_cast<double>(t)));
}

constexpr Value<Scale::Celcius> operator"" _deg(long double t) { return Value<Scale::Celcius>(t); }
constexpr Value<Scale::Kelvin> operator"" _K(long double t) { return Value<Scale::Kelvin>(t); }
constexpr Value<Scale::Fahrenheit> operator"" _f(long double t)
{
    return Value<Scale::Fahrenheit>(t);
}
constexpr Value<Scale::Celcius> operator"" _deg(unsigned long long t)
{
    return Value<Scale::Celcius>(t);
}
constexpr Value<Scale::Kelvin> operator"" _K(unsigned long long t)
{
    return Value<Scale::Kelvin>(t);
}
constexpr Value<Scale::Fahrenheit> operator"" _f(unsigned long long t)
{
    return Value<Scale::Fahrenheit>(t);
}
} // namespace Temperature

int main()
{
    using namespace Temperature;
    auto t1 = -10_deg;
    auto t2 = -10_f;

    auto t3 = cast_scale<Scale::Fahrenheit>(t1);
    std::cout << t1 << std::endl;
    std::cout << t2 << std::endl;
    std::cout << t3 << std::endl;
    return 0;
}